### DISCUSSION TITLE
Modelling question - How to bend cylinder to get tube

### DISCUSSION THREAD

**Discussion by jaroslavknotek** [mitsuba-renderer/mitsuba3#277] [Q&A] [answered]
How can I create bent tube within mitsuba easily? 

As an input, I have a curve that I don't know in advace (therefore I can't model it outside mitsuba) but I know that it's bent just slightly (as you can in the picture below) and it start point is always at origin [0,0,0] and end is at [0,0,1].
![bent](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2F3dwarehouse.sketchup.com%2Fwarehouse%2Fv1.0%2Fpubliccontent%2F4b64a9cd-12c2-42ed-aa21-9cdf7a3b3cf2&f=1&nofb=1)

I was wodering what is the quickest way to come up with such tube within mitusba. I looked around a bit and found out that
- You can create your own mesh by following [this tutorial](https://github.com/mitsuba-renderer/mitsuba-tutorials/blob/master/how_to_guides/mesh_io_and_manipulation.ipynb)
- You can't manipulate with mitsuba's [cylinder](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_shapes.html#cylinder-cylinder) 
- You can import mesh from outside and then manipulate it.

Based on the assumptions and observations, I came up with this approach
- I will create a template of a cylinder that will be composed of n segments
- Each segment will be shifted in X,Y axis acording to the position of respective point of the curve. 

Given the assumption about Z-orientation and small curvature, I can approximate the tube well enough in this way without significant anomalies. The template itself can be either pre-created mesh or something created in memory.

Is there any quicker way in mitsuba to achieve this? Or any other tips (e.g. way to split bend Cylinder shape or compose the bent  from small straight pieces in some continuous way)

Thank you

**Comment by njroussel** [comment, marked as answer]
I believe everything you mentioned is correct.

However, I'm not completely sure of what you mean by "template of a cylinder". If you're going to try to stitch together an existing mesh model that seems a bit painful and you might as well do it directly in Mitsuba using the mesh manipulation tools.
Here's what I would suggest:

1. Place N points on the curve such that subsequent points are at equal distances from one another (including state/end points of the curve).
2. For every point generate D vertices for a circle. The circle is centered on the point and is directed along the curve. These vertices will be the final vertices of the cylinder
3. "Connect" subsequent circles by specifying the faces between them. This step effectively creates the faces of the cylinder. 

The tutorial you linked already has similar steps, you can easily adapt it to this use case.

Although this is all possible with Mitsuba/Dr.Jit, you might be better off using a different set of tools that are specifically designed for this and then import it into Mitsuba.

**Reply by jaroslavknotek** [reply]
Thank you. Your answer is similar to what I meant. In my case, I created the cylinder mesh with D connected parts and then I moved their center to correspond to the points of the curve.

In the end, it was much easier to create this in numpy than import 3rd party libraries that does the same thing properly. I tried to lookup libraries that offer "extrude along path" functionality and found two:
- [pygimil](https://www.pygimli.org/_examples_auto/1_meshing/plot_extrude_2D_to_3D.html)
- [salome](https://docs.salome-platform.org/latest/gui/SMESH/tui_modifying_meshes.html#extrusion-along-a-path)
The first one is not easy to install, since it requires conda (which I don't use for this project) and the other is clunky.
